### PR TITLE
Support recording data from multiple ADXL345's in one run, and more.

### PR DESCRIPTION
**Pull request by miklschmidt**
This PR adds `CHIP` and `POINT` arguments to `TEST_RESONANCES`. It was born out of a need to script generation of belt tension graphs for the upcoming V-Core IDEX printer, without having to edit the config between each test.

Example:

```properties
[gcode_macro GENERATE_IDEX_GRAPHS]
gcode:
  {% if params.AXIS is defined %}
      {% if params.AXIS|lower == 'x' %}
          G28
          TEST_RESONANCES AXIS=X CHIP="adxl345" OUTPUT=raw_data NAME=t0
          T1
          TEST_RESONANCES AXIS=X CHIP="rpi" OUTPUT=raw_data NAME=t1
          G4 P5000
          RUN_SHELL_COMMAND CMD=generate_tension_graphs PARAMS=x
          RESPOND MSG="Belt tension graph generated for the X axis. You'll find it in the input_shaper folder in the machine tab!"
      {% elif params.AXIS|lower == 'y' %}
          G28
          TEST_RESONANCES AXIS=Y CHIP="adxl345, rpi" OUTPUT=raw_data NAME=belt_tension POINT={printer.toolhead.axis_maximum.x},{printer.toolhead.axis_maximum.y/2},20
          G4 P5000
          RUN_SHELL_COMMAND CMD=generate_tension_graphs PARAMS=y
          RESPOND MSG="Belt tension graph generated for the Y axis. You'll find it in the input_shaper folder in the machine tab!"
      {% else %}
          {action_raise_error("Unknown axis specified. Expected X or Y.")}
      {% endif %}
  {% else %}
      G28
      TEST_RESONANCES AXIS=Y CHIP="adxl345, rpi" OUTPUT=raw_data NAME=belt_tension POINT={printer.toolhead.axis_maximum.x},{printer.toolhead.axis_maximum.y/2},20
      TEST_RESONANCES AXIS=X CHIP="adxl345" OUTPUT=raw_data NAME=t0
      T1
      TEST_RESONANCES AXIS=X CHIP="rpi" OUTPUT=raw_data NAME=t1
      G4 P5000
      RUN_SHELL_COMMAND CMD=generate_tension_graphs PARAMS=y
      G4 P5000
      RUN_SHELL_COMMAND CMD=generate_tension_graphs PARAMS=x
      RESPOND MSG="Belt tension graph generated for X and Y. You'll find them in the input_shaper folder in the machine tab!"
  {% endif %}
  T0
  M84
  ```
  
It has been very helpful while trying to get the belt tension right on the upcoming Rat Rig V-Core IDEX (and i would imagine the Voron Tridex as well), and i expect to ship it with RatOS once the printer is released, so here's me getting ahead of that 😅

### CODE REVIEW

**Review by ld3300 — COMMENTED** [review]
Hi, I can't comment on the validity of the code, but please add the [signed-off-by:](https://www.klipper3d.org/CONTRIBUTING.html#format-of-commit-messages) statements.

In docs/G-Codes.md please also adjust your line lengths to stay within the desired 80 character columns per line.
Best,
-David

**Comment by miklschmidt** [comment]
@ld3300 done, thank you :)

**Comment by dmbutyugin** [comment]
Thanks!

I first had a high-level question: could you please elaborate how the accelerometer setup will look like and how the data is used for belt tensioning then? I imagine the second accelerometer isn't located on the bed?

BTW, I do think it might be a really good idea to let a user override a test point(s). For the accelerometer, does a test like `TEST_RESONANCES AXIS=Y CHIP="adxl345, rpi" OUTPUT=raw_data` provide useful info (vs just using one accelerometer)? As an alternative to your change, maybe for IDEX printers it would make sense to rework the configuration of accelerometer chips in `[resonance_tester]`? Because the latter was initially made without IDEX printers in mind.

**Comment by miklschmidt** [comment]
> could you please elaborate how the accelerometer setup will look like and how the data is used for belt tensioning then? I imagine the second accelerometer isn't located on the bed?

Yes! So one accelerometer is mounted on each toolhead (T0 and T1), when using it to measure the Y belts on the Trident IDEX for example, each toolhead is placed in it's inactive position (as close to the X/Y joiner as possible), and data is read from both accelerometers, there's almost zero cross contimination as far as i can tell, so you can combine the results from both to see how the resonance of your two Y belts compares to each other, same way it's done on CoreXY. For the individual belt controlling the X axis of the toolheads, you measure the belts by testing the X axis on one, and then the other, separately.

> For the accelerometer, does a test like TEST_RESONANCES AXIS=Y CHIP="adxl345, rpi" OUTPUT=raw_data provide useful info (vs just using one accelerometer)?

Yes, separating the resonances from the two Y belts would be hard without it (as neither carriage can reach the opposite side of the gantry, and the further towards the center you go, the more "contamination" you'd have from the other belt). I'm not sure it's useful for other setups than IDEX, though. I guess it could be used on a bed slinger, but mostly just to cut time (ie. recording from X and Y accelerometers at the same time and combining them in the input shaper graphs).

> As an alternative to your change, maybe for IDEX printers it would make sense to rework the configuration of accelerometer chips in [resonance_tester]?

Yes, if not as an alternative, it would be nice as a supplement too.

File: klippy/extras/resonance_tester.py [hunk 1]
```diff
@@ -299,18 +334,22 @@ def cmd_MEASURE_AXES_NOISE(self, gcmd):
...
         return name_suffix.replace('-', '').replace('_', '').isalnum()
 
-    def get_filename(self, base, name_suffix, axis=None, point=None):
+    def get_filename(self, base, name_suffix, axis=None,
+                     point=None, chip_name=None):
         name = base
         if axis:
             name += '_' + axis.get_name()
+        if chip_name:
+            name += '_' + chip_name
```
Comment (dmbutyugin) [inline klippy/extras/resonance_tester.py:343]: chip_name may contain spaces (e.g. be "adxl345 toolhead1"). Ideally, you'd want to replace them with something, e.g. call replace(" ", "_")

**Review by dmbutyugin — COMMENTED** [review]
Thanks a lot for the explanations!

> >    For the accelerometer, does a test like TEST_RESONANCES AXIS=Y CHIP="adxl345, rpi" OUTPUT=raw_data provide useful info (vs just using one accelerometer)?

> Yes, separating the resonances from the two Y belts would be hard without it (as neither carriage can reach the opposite side of the gantry, and the further towards the center you go, the more "contamination" you'd have from the other belt). I'm not sure it's useful for other setups than IDEX, though. I guess it could be used on a bed slinger, but mostly just to cut time (ie. recording from X and Y accelerometers at the same time and combining them in the input shaper graphs, in case Y accel is affecting the X gantry, for example on cantilevered printers like the V-Minion).

TBH, in order to avoid potential complications in the future with new test resonance methods, where we may want to combine the signals from multiple accelerometers, I would slightly prefer that we simplified this change a bit and only allow a single chip to be specified in `TEST_RESONANCES CHIP` command, even if at the expense of longer test times (33% time increase AFAICT). But this is merely a preference, and not a blocker.

**Comment by dmbutyugin** [comment]
Separately, but on a topic of IDEX printers, I've made a PR #5311 so that the input shaper parameters could be specified for the dual carriage of an IDEX printer independenly as (e.g. for X carriage)
```
[input_shaper]
shaper_type_dc: <dual_carriage_x_shaper>
shaper_freq_dc: <dual_carriage_x_freq>
```
If you are interested in it and happen to have a hardware, I'd encourage you to give it a try :)

**Comment by miklschmidt** [comment]
> TBH, in order to avoid potential complications in the future with new test resonance methods, where we may want to combine the signals from multiple accelerometers, I would slightly prefer that we simplified this change a bit and only allow a single chip to be specified in TEST_RESONANCES CHIP command, even if at the expense of longer test times (33% time increase AFAICT). But this is merely a preference, and not a blocker.

I understand, that makes sense. 33% is *a lot* though, especially when accumulated over 10-15 runs trying to get the tension *just right*. It's already a process that takes hours, so preferably we could come up with some way to still record data from both that wouldn't cause problems down the line? A new command maybe?

File: klippy/extras/resonance_tester.py [hunk 2]
```diff
@@ -165,29 +171,36 @@ def _run_test(self, gcmd, axes, helper, raw_name_suffix=None):
...
                 if helper is None:
                     continue
-                for chip_axis, aclient in raw_values:
+                for chip_axis, aclient, chip_name in raw_values:
                     if not aclient.has_valid_samples():
                         raise gcmd.error(
-                                "%s-axis accelerometer measured no data" % (
-                                    chip_axis,))
+                            "%s-axis accelerometer '%s' measured no data" % (
+                                chip_axis,chip_name,))
```
[the remark this answers is not in the document]
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py:203]: What if you have multiple chips for the same axis? Which is the case on Y for non-cartesian IDEX machines. (sorry i'm being a little vague on the kinematics here, the V-Core IDEX is still under NDA)

File: klippy/extras/resonance_tester.py [hunk 3]
```diff
@@ -198,6 +211,24 @@ def _run_test(self, gcmd, axes, helper, raw_name_suffix=None):
...
         # Parse parameters
         axis = _parse_axis(gcmd, gcmd.get("AXIS").lower())
+        accel_chips = gcmd.get("CHIP", None)
+        test_point = gcmd.get("POINT", None)
+
+        if test_point:
+            test_coords = test_point.split(',')
+            if len(test_coords) != 3:
+                raise gcmd.error("Invalid POINT parameter, must be 'x,y,z'")
+            test_point = [float(p.strip()) for p in test_coords]
```
[the remark this answers is not in the document]
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py:221]: Good question.
```
23:35:34
[TEST_RESONANCES AXIS=Y POINT=lol]()
23:35:34
Invalid POINT parameter, must be 'x,y,z'
```

Seems fine :)

File: klippy/extras/resonance_tester.py [hunk 4]
```diff
@@ -287,7 +321,8 @@ def cmd_MEASURE_AXES_NOISE(self, gcmd):
         for chip_axis, aclient in raw_values:
             if not aclient.has_valid_samples():
                 raise gcmd.error(
-                        "%s-axis accelerometer measured no data" % (chip_axis,))
+                        "%s-axis accelerometer measured no data" % (
+                            chip_axis,))
```
[the remark this answers is not in the document]
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py:325]: It's needed to stay within the 80 character line width limit

File: klippy/extras/resonance_tester.py [hunk 5]
```diff
@@ -198,6 +211,24 @@ def _run_test(self, gcmd, axes, helper, raw_name_suffix=None):
     def cmd_TEST_RESONANCES(self, gcmd):
         # Parse parameters
         axis = _parse_axis(gcmd, gcmd.get("AXIS").lower())
+        accel_chips = gcmd.get("CHIP", None)
```
[the remark this answers is not in the document]
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py:214]: Fixed 👍

**Comment by dmbutyugin** [comment]
> > TBH, in order to avoid potential complications in the future with new test resonance methods, where we may want to combine the signals from multiple accelerometers, I would slightly prefer that we simplified this change a bit and only allow a single chip to be specified in TEST_RESONANCES CHIP command, even if at the expense of longer test times (33% time increase AFAICT). But this is merely a preference, and not a blocker.
> 
> I understand, that makes sense. 33% is _a lot_ though, especially when accumulated over 10-15 runs trying to get the tension _just right_. It's already a process that takes hours, so preferably we could come up with some way to still record data from both that wouldn't cause problems down the line? A new command maybe?

Pardon me for probably a very naive question, but why not check belts tension with a frequency analyzer app on the phone? Are there any considerations regarding that approach?

File: klippy/extras/resonance_tester.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py]: Doh, of course you're right, fixed!

File: klippy/extras/resonance_tester.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py]: > Sorry for not being specific, I had a different test case like `TEST_RESONANCES AXIS=Y POINT=x,y,z` in mind

Honestly that one's on me, might have been a little quick there. I've added a proper check 👍

**Comment by miklschmidt** [comment]
> > > TBH, in order to avoid potential complications in the future with new test resonance methods, where we may want to combine the signals from multiple accelerometers, I would slightly prefer that we simplified this change a bit and only allow a single chip to be specified in TEST_RESONANCES CHIP command, even if at the expense of longer test times (33% time increase AFAICT). But this is merely a preference, and not a blocker.
> > 
> > 
> > I understand, that makes sense. 33% is _a lot_ though, especially when accumulated over 10-15 runs trying to get the tension _just right_. It's already a process that takes hours, so preferably we could come up with some way to still record data from both that wouldn't cause problems down the line? A new command maybe?
> 
> Pardon me for probably a very naive question, but why not check belts tension with a frequency analyzer app on the phone? Are there any considerations regarding that approach?

It's a good question, and to be honest, it's probably "good enough" (provided people can agree what the frequency should actually be), but we've been spoiled with all the juicy details provided by the comparison graphs as layed out here: https://www.klipper3d.org/Measuring_Resonances.html#testing-custom-axes

Most people prefer using the adxl345 to make sure their tension is as even as possible, a frequency tuner app won't tell you about multiple peaks, different amplitudes etc. It's just a lot more detailed analysis, and 3d printing enthusiasts tend to be very perfectionist :D

File: klippy/extras/resonance_tester.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmbutyugin) [reply klippy/extras/resonance_tester.py:214]: I'm sorry, I see that you've updated the documentation to `CHIPS`, but the code here still uses `CHIP` parameter.

File: klippy/extras/resonance_tester.py [hunk 6]
```diff
@@ -198,6 +211,28 @@ def _run_test(self, gcmd, axes, helper, raw_name_suffix=None):
...
+            if len(test_coords) != 3:
+                raise gcmd.error("Invalid POINT parameter, must be 'x,y,z'")
+            try:
+                test_point = [float(p.strip()) for p in test_coords]
+            except ValueError:
+                raise gcmd.error("Invalid POINT parameter, must be 'x,y,z'"
+                " where x, y and z are valid floating point numbers")
+
+        if accel_chips:
+            initialized_chips = []
```
Comment (dmbutyugin) [inline klippy/extras/resonance_tester.py:228]: Maybe `parsed_chips` or something like that. 'Initialized' implies that some sort of initialization was performed, but it is not the case.

**Comment by dmbutyugin** [comment]
> It's a good question, and to be honest, it's probably "good enough" (provided people can agree what the frequency should actually be), but we've been spoiled with all the juicy details provided by the comparison graphs as layed out here: https://www.klipper3d.org/Measuring_Resonances.html#testing-custom-axes
> 
> Most people prefer using the adxl345 to make sure their tension is as even as possible, a frequency tuner app won't tell you about multiple peaks, different amplitudes etc. It's just a lot more detailed analysis, and 3d printing enthusiasts tend to be very perfectionist :D
> 
> I actually did try tuning this by ear, it did not go well.

Yes, I wouldn't think that tuning by ear would be precise enough. I was just curious if you had experience with any frequency analysis app on the phone and how it fairs against this approach. I think it would show multiple peaks, but I agree that it may be not precise enough. Well, if there is interest in using this test for belt tension adjustments, it might be good to merge it to Klipper. Though if different methods will be used in the future for resonance testing, I'm not sure how suitable they would be for the belt tension tests. But we can see how it'll be later.

File: klippy/extras/resonance_tester.py (hunk 6 above)
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py]: fixed 👍

File: klippy/extras/resonance_tester.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (miklschmidt) [reply klippy/extras/resonance_tester.py]: Fixed 👍

**Comment by miklschmidt** [comment]
Sorry for the delay! Got buried in my inbox :D

**Review by dmbutyugin — APPROVED** [review]
OK, thank you, looks good now! And sorry for the delay with the review.

**Comment by KevinOConnor** [comment]
Thanks.

-Kevin

**Pull request merged**